buku

```diff
@@ -1841,28 +1850,34 @@ class BukuDb:
         delay_commit : bool
             True if record should not be committed to the DB,
             leaving commit responsibility to caller. Default is False.
+        upto : int, optional
+            If specified, multiple indices are moved at once.
+        retain_order: bool
+            Shift indices of multiple records by 1 instead of replacing
+            the deleted record with the last one. Default is False.
         """
 
         # Return if the last index left in DB was just deleted
         max_id = self.get_max_id()
-        if not max_id:
+        if not max_id or (upto and upto < index):
             return
 
-        query1 = 'SELECT id, URL, metadata, tags, desc, flags FROM bookmarks WHERE id = ? LIMIT 1'
-        query2 = 'DELETE FROM bookmarks WHERE id = ?'
-        query3 = 'INSERT INTO bookmarks(id, URL, metadata, tags, desc, flags) VALUES (?, ?, ?, ?, ?, ?)'
-
         # NOOP if the just deleted index was the last one
         if max_id > index:
-            results = self._fetch(query1, max_id)
-            for row in results:
-                with self.lock:
-                    self.cur.execute(query2, (row.id,))
-                    self.cur.execute(query3, (index, row.url, row.title, row.tags_raw, row.desc, row.flags))
-                    if not delay_commit:
-                        self.conn.commit()
-                if self.chatty:
-                    print('Index %d moved to %d' % (row.id, index))
+            with self.lock:
+                if retain_order or (upto or 0) > index:
+                    step = (max(max_id - upto, upto + 1 - index) if not retain_order else
+                            1 if not upto else upto + 1 - index)
```

Comment:
There's 3 possibility here:
* retain-order without upto
* retain-order with upto
* upto without retain-order

The 3rd case only differs from the 2nd one in that we want to support the sub-case where there's more records with larger indices remaining than was in the deleted range ("Indices 9998-10000 moved to 9990-9992"). This is covered by the `max()` subexpression.